tests/webTables.spec.ts

```diff
@@ -1,222 +1,133 @@
+        /* 2. In the list of Owners, locate the owner by the name "Jeff Black". 
+        Add the assertions that this owner is from the city of "Monona" and he has a pet with a name "Lucky" */
+        // Find Jeff Black in the table 
+        const row = page.getByRole('row', {name: 'Jeff Black'})
+    
+        // Validate the owner is from Monona
+        await expect(row.locator('td').nth(2)).toHaveText('Monona')
+    
+        // Validate the owner has a pet named Lucky
+        await expect(row.locator('td').nth(4)).toHaveText('Lucky')
+    
+    });
+    
+    test('Validate owners count of the Madison city', async ({ page }) => {
+        /* 2. In the list of Owners, locate all owners who live in the city of "Madison". 
+        Add the assertion that the total number of owners should be 4 */
+        
+        // Validate there are only 4 entries of 'Madison' in the city column
+        await expect(page.getByRole('row', {name: 'Madison'})
+            // Filtering specifically for the City column
+            .filter({has: page.getByRole('cell').nth(2)})).toHaveCount(4)
+    
+    });
+    
+    test('Validate search by Last Name', async ({ page }) => {
+        /* 2. - 8. In the "Last name" input field, type the last name "Black", "Davis", "Es" (partial string), and "Playwright". */
+        // Search for the following last names and validate the results
+        const lastNames = ['Black', 'Davis', 'Es', 'Playwright']
+    
+        for(const lastName of lastNames){
+            // Use the search field to locate entries with a specific last name
+            await page.getByRole('textbox').fill(lastName)
+            await page.getByText('Find Owner').click()
+            
+            // Capture the rows in the table body (not the header)
+            const rows = await page.locator('tbody tr').all()
+    
+            // For each resulting row, check for the last name
+            for(const row of rows){
+    
+                // Ensuring the resulting row has cells. There is an extra row in the 'Pets' column that we need to filter out, as it contains no cells.
+                if((await row.getByRole('cell').all()).length > 0){
+    
+                    // Checking if the name isn't our negative test, 'Playwright'.
+                    if(lastName !== 'Playwright'){
+    
+                        // Validating that the last name is in the first cell in the row
+                        await expect(row.getByRole('cell').first()).toContainText(lastName) 
+                    }
+                    else{
+                        /* 9. Add the assertion of the message "No owners with LastName starting with "Playwright"" */
+                        // Validate that the error message is correct for a last name that cannot be found
+                        await expect(page.locator('h2').locator('..').locator('div').last()).toHaveText("No owners with LastName starting with \"Playwright\"")
```

Comment:
Simplify this locator.
**I do NOT allow** you to use `the locator('..')` anymore :)))) It looks like you are abusing this technique too much making the locators overcomplicated. It's all over the place.

Don't start with children element going up to parent. Start with patent and go down. And use as many nodes in your locator as possible.

Reply:
You got it!  I accept this challenge.